fdbclient/GlobalConfig.actor.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * GlobalConfig.actor.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#if defined(NO_INTELLISENSE) && !defined(FDBCLIENT_GLOBALCONFIG_ACTOR_G_H)
+#define FDBCLIENT_GLOBALCONFIG_ACTOR_G_H
+#include "fdbclient/GlobalConfig.actor.g.h"
+#elif !defined(FDBCLIENT_GLOBALCONFIG_ACTOR_H)
+#define FDBCLIENT_GLOBALCONFIG_ACTOR_H
+
+#include <any>
+#include <map>
+#include <unordered_map>
+
+#include "fdbclient/CommitProxyInterface.h"
+#include "fdbclient/ReadYourWrites.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+
+// The global configuration is a series of typed key-value pairs synced to all
+// nodes (server and client) in an FDB cluster in an eventually consistent
+// manner.
+
+class GlobalConfig {
+public:
+	GlobalConfig();
+	GlobalConfig(const GlobalConfig&) = delete;
+	GlobalConfig& operator=(const GlobalConfig&) = delete;
+
+	static void create(DatabaseContext* cx, Reference<AsyncVar<ClientDBInfo>> dbInfo);
+	static GlobalConfig& globalConfig();
+
+	const std::any get(KeyRef name);
+	const std::map<KeyRef, std::any> get(KeyRangeRef range);
+
+	// To write into the global configuration, submit a transaction to
+	// \xff\xff/global_config/<your-key> with <your-value> encoded using the
+	// FDB tuple typecodes.
+
+	Future<Void> onInitialized();
+
+private:
+	void insert(KeyRef key, ValueRef value);
+	void erase(KeyRef key);
+	void erase(KeyRangeRef range);
+
+	ACTOR static Future<Void> refresh(GlobalConfig* self);
+	ACTOR static Future<Void> updater(GlobalConfig* self, Reference<AsyncVar<ClientDBInfo>> dbInfo);
+
+	Database cx;
+	Future<Void> _updater;
+	Promise<Void> initialized;
+	Arena arena;
+	std::unordered_map<StringRef, std::any> data;
```

Comment:
I think one way of solving your memory problem is to change the value type of this map to a new type that looks like this:

```
struct ConfValue {
    Arena arena;
    std::any value;
};
```

Your range query could still return `std::map<KeyRef, std::any>` but then you'll have to be very careful with object lifetime (you could also create a new type and reference count the ranges you give out -- or you could just also return the Arena as they're reference counted as well)

Reply:
I added an initial attempt in https://github.com/apple/foundationdb/pull/4330/commits/b22e91ddbabeaccf76ecdc41622c9219fd111a2e. I don't like making the API more complicated than necessary, so for types that match `std::is_arithmetic` I added a separate `get` function which bypasses the arena and just returns a copy of the object. For regular objects, the API now returns the `ConfigValue` object which means as long as the caller retains the return value, they should be able to read the value.